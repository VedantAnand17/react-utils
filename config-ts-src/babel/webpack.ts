import { pick } from 'lodash';

import {
  type PluginOptionsT as ReactCssModulesOptionsT,
} from '@dr.pogodin/babel-plugin-react-css-modules';

const {
  generateScopedNameFactory,
} = require('@dr.pogodin/babel-plugin-react-css-modules/utils');

const generateScopedNameDev = generateScopedNameFactory(
  '[package]___[path][name]___[local]___[hash:base64:6]',
);

const generateScopedNameProd = generateScopedNameFactory(
  '[hash:base64:6]',
);

export interface BabelCompilerI {
  env: () => string;
}

export type OptionsT = {
  noRR?: boolean;
  noStyling?: boolean;
  targets?: string | string[] | { [key: string]: string };
};

/**
 * Supported Babel environments.
 */
export enum ENVIRONMENTS {
  DEV = 'development',
  PROD = 'production',
  TEST = 'test',
}

export interface PresetOrPluginOptionsI {}

type PresetOrPluginT = string | [string, PresetOrPluginOptionsI];

export type ConfigT = {
  presets: PresetOrPluginT[];
  plugins: PresetOrPluginT[];
};

/**
 * Creates a new base config.
 * @param [options] Base config options.
 * @return Generated config.
 */
function newBaseConfig(options: OptionsT = {}): ConfigT {
  return {
    presets: [
      // Chrome 69 is the browser for Android API 28.
      ['@babel/env', { targets: options.targets || 'defaults or chrome >= 69' }],

      // TODO: Starting from Babel 8, "automatic" will be the default runtime,
      // thus once upgraded to Babel 8, runtime should be removed from
      // @babel/react options below.
      ['@babel/react', { runtime: 'automatic' }],

      '@dr.pogodin/babel-preset-svgr',

      // TODO: Should we rather optionally include it for TS projects only?
      '@babel/typescript',
    ],
    plugins: [
      ['module-resolver', {
        extensions: ['.js', '.jsx', '.ts', '.tsx'],
        root: [
          './src/shared',
          './src',
        ],
      }],
      '@babel/transform-runtime',
    ],
  };
}

/**
 * Updates given `config` with styling ((S)CSS-related) setup.
 *
 * **Beware:** It mutates `config`.
 *
 * @param {object} config
 * @param {string} env Target environment: `development` or `production`.
 * @return {object} Returns mutated config for chaining.
 * @ignore
 */
function addStyling(config: ConfigT, env: ENVIRONMENTS) {
  const cssModulesOps: ReactCssModulesOptionsT = {
    autoResolveMultipleImports: true,
    filetypes: {
      '.scss': { syntax: 'postcss-scss' },
    },
  };
  config.plugins.push(['@dr.pogodin/react-css-modules', cssModulesOps]);
  switch (env) {
    case ENVIRONMENTS.DEV:
    case ENVIRONMENTS.TEST:
      cssModulesOps.generateScopedName = generateScopedNameDev;
      break;
    case ENVIRONMENTS.PROD:
      cssModulesOps.generateScopedName = generateScopedNameProd;
      break;
    default:
  }
  return config;
}

/**
 * Generates Babel preset for Webpack.
 * @param babel Babel compiler.
 * @param [ops] Preset options.
 * @param [ops.noRR] If truthy `react-refresh/babel` plugin is not
 *  included into config, no matter the environment.
 * @param [ops.noStyling] If truthy all setup related to styling
 *  ((S)CSS processing) will be skipped.
 * @param [ops.targets=defaults] Targets for
 *  `@babel/preset-env`.
 * @return Generated config.
 */
export default function getPreset(babel: BabelCompilerI, ops: OptionsT = {}) {
  const env = babel.env();

  const baseOps = pick(ops, ['targets']);
  const res = newBaseConfig(baseOps);

  if (!ops.noStyling) addStyling(res, env as ENVIRONMENTS);
  if (env === ENVIRONMENTS.DEV && !ops.noRR) {
    res.plugins.push('react-refresh/babel');
  }
  return res;
}
