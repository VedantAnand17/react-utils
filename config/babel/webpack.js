/* Babel preset for the Webpack build. */

const _ = require('lodash');

const config = {
  presets: [
    '@babel/env',
    '@babel/react',
  ],
  plugins: [
    ['module-resolver', {
      extensions: ['.js', '.jsx'],
      root: [
        './src/shared',
        './src',
      ],
    }],
    '@babel/syntax-dynamic-import',
    'inline-react-svg',
    '@babel/transform-runtime',
    ['react-css-modules', {
      filetypes: {
        '.scss': {
          syntax: 'postcss-scss',
        },
      },
    }],
  ],
};

function getPreset(babel) {
  const res = _.cloneDeep(config);

  const reactCssModulePluginOptions = res.plugins.find(
    ([name]) => name === 'react-css-modules',
  )[1];

  switch (babel.getEnv()) {
    case 'development':
      reactCssModulePluginOptions
        .generateScopedName = '[path][name]___[local]___[hash:base64:6]';
      res.plugins.push('react-hot-loader/babel');
      break;
    case 'production':
      reactCssModulePluginOptions.generateScopedName = '[hash:base64:6]';
      break;
    default:
  }
  return res;
}

module.exports = getPreset;
