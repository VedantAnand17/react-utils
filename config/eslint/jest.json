{
  "env": {
    "browser": true,
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "@typescript-eslint/no-implied-eval": 0,
    "global-require": 0,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0,
    "no-new-func": 0,
    "react/require-default-props": 0
  }
}