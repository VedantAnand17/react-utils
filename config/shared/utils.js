/**
 * Misc utilities reusable in different JS-format config files.
 */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const path = require('path');
const cssesc = require('cssesc');

const { interpolateName } = require('loader-utils');

/**
 * Normalizes file path to OS-independent format (adopted from css-loader).
 * @ignore
 * @param {string} file
 * @return {string}
 */
function normalizePath(file) {
  return path.sep === '\\' ? file.replace(/\\/g, '/') : file;
}

// eslint-disable-next-line no-control-regex
const filenameReservedRegex = /[<>:"/\\|?*]/g;
// eslint-disable-next-line no-control-regex
const reControlChars = /[\u0000-\u001f\u0080-\u009f]/g;

function escapeLocalident(localident) {
  return cssesc(
    localident
      // For `[hash]` placeholder
      .replace(/^((-?[0-9])|--)/, '_$1')
      .replace(filenameReservedRegex, '-')
      .replace(reControlChars, '-')
      .replace(/\./g, '-'),
    { isIdentifier: true },
  );
}

/**
 * Returns the name of package containing the folder; i.e. it recursively looks
 * up from the folder for the closest package.json file, and returns the name in
 * that file. It also caches the results from previously fisited folders.
 * @ignore
 * @param {string} folder
 * @return {string}
 */
function getPackageInfo(folder) {
  let res = getPackageInfo.cache[folder];
  if (!res) {
    const p = path.resolve(folder, 'package.json');
    /* eslint-disable global-require, import/no-dynamic-require */
    res = fs.existsSync(p) ? {
      name: require(p).name,
      root: folder,
    } : getPackageInfo(path.resolve(folder, '..'));
    /* eslint-enable global-require, import/no-dynamic-require */
    getPackageInfo.cache[folder] = res;
  }
  return res;
}
getPackageInfo.cache = {};

function getLocalIdent(
  { resourcePath },
  localIdentName,
  localName,
  options = {},
) {
  const packageInfo = getPackageInfo(path.dirname(resourcePath));
  const request = normalizePath(path.relative(packageInfo.root, resourcePath));
  return interpolateName({
    resourcePath,
  }, localIdentName, {
    ...options,
    content: `${packageInfo.name + request}\x00${localName}`,
    context: packageInfo.root,
  }).replace(/\[package\]/gi, packageInfo.name)
    .replace(/\[local\]/gi, localName)
    .replace(/@/g, '-');
}

function generateScopedNameDev(localName, assetPath) {
  // Mind: "css-loader" additionally escapes identifiers returned by
  // getLocalIdent(); the babel plugins - don't, thus we need to escape
  // ourselves.
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[package]___[path][name]___[local]___[hash:base64:6]',
      localName,
      {},
    ),
  );
}

function generateScopedNameProd(localName, assetPath) {
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[hash:base64:6]',
      localName,
      {},
    ),
  );
}

module.exports = {
  generateScopedNameDev,
  generateScopedNameProd,
  getLocalIdent,
};
