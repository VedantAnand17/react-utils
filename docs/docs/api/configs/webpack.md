# Webpack
The library provides several tuneable [Webpack] configs, intended for different
kinds of target code (an [application] or a [library]), and environments
(development or production).

Common for every config variant, the library exposes it as a "config factory
function", which generates the actual configuration once called with necessary
options. That config can be further customized by the host code, and then passed
into [Webpack].

For example, one can use the production app config the following way.
In the host codebase the `/config/webpack/production.js` file is created:
```js title="/config/webpack/production.js"
const { merge } = require('webpack-merge');
const configFactory = require(
  '@dr.pogodin/react-utils/config/webpack/app-production',
);

const baseConfig = configFactory({
  context: '/path/to/context/folder',
  entry: 'relative/path/to/entry/point',
});

module.exports = merge(baseConfig, {
  // Additional Webpack config to merge with the base one.
});
```

Then in the main `webpack.config.js` file do:
```js title="/webpack.config.js"
module.exports = function buildConfig(env) {
  const config = require(`./config/webpack/${env}.js`);
  // Some additional operations on config may happen here.
  return config;
}
```

## Application

### Base {#app-base}
```js
const configFactory = require(
  '@dr.pogodin/react-utils/config/webpack/app-base',
);
```
This factory **function configFactory(options)** &rArr; **object** creates
the base [Webpack] config for apps, which includes the setup common for both
[development](#app-dev) and [production](#app-prod) environments. The `options`
argument is an object with the following valid fields:

- **Required**:
  - `babelEnv` - **string** - Environment to use for [Babel] compilation.
  - `context` - **string** - Base URL for relative paths resolution.
  - `entry` - **string | string[]** - Entry point(s) for the "main" chunk(s).
    Under the hood the config will prepend them by some polyfills, like
    [babel-polyfill](https://babeljs.io/docs/usage/polyfill/),
    [nodelist-foreach-polyfill](https://www.npmjs.com/package/nodelist-foreach-polyfill),
    _etc._
  - `mode` - **string** - [Webpack mode](https://webpack.js.org/concepts/mode/).
  - `publicPath` - **string** - Base URL for the output of build assets.
- **Optional**:
  - `babelLoaderOptions` - **object** - Overrides default [Babel] loader options
    for JSX and SVG files.
  - `cssLocalIdent` - **string** - Template for CSS classnames generated by
    the [Webpack]'s `css-loader`. The value is passed into the `localIdentName`
    option of the loader. It should match corresponding option in the [Babel]
    config. Defaults `hash:base64:6`.
  - `dontEmitBuildInfo` - **boolean** - Opts out of creation of `.build-info`
    file in the codebase root during the compilation.
  - `workbox` - **boolean | object** - When set the
    [Workbox's InjectManifest plugin](https://developers.google.com/web/tools/workbox/modules/workbox-webpack-plugin#injectmanifest_plugin)
    is included into the array of [Webpack] plugins, and it generates server
    worker for browser. If the value is an object it is merged into options
    passed into the plugin, otherwise default options are used:
    ```json
    {
      "importWorkboxFrom": "local",
      "swSrc": "@dr.pogodin/react-utils/config/workbox/default.js",
      "swDest": "__service-worker.js"
    }
    ```

    If the service worker is generated by this option, it will be automatically
    initiated at the client side by the standard
    [client-side initialization script](docs/api/utils/client)
    provided by this library.
    
    Note that `swDest`'s value cannot be overriden by config options provided
    via `workbox` object.

  - `keepBuildInfo` - **boolean** - If set and a `.build-info` file from
    a previous build exists in the context directory, it will be loaded and
    used, rather than re-generated by the config factory. It allows to re-create
    the Webpack config during a server launch without re-generation of the build
    info file created during a previous build (and thus bundled into the
    frontend bundle).

  - `outputPath` - **string** - Output build path, relative to the `context`.
    Defaults `build`.

  - `sitemap` - **function | string** - The path to JS or JSON config for
    sitemap. It can be relative to the context, and can be a factory, which
    returns the config. The config should be compatible with
    [`sitemap`](https://www.npmjs.com/package/sitemap) library, and if
    provided the Webpack config factory will use it to gererate `sitemap.xml`
    file in the output folder, and then serve it from the app root.

The config factory returns a config object which
- Bundles font assets (EOF, OTF, SVG, TTF, WOFF, WOFF2 files)
  from the `src/assets/fonts` folder of source code and emits them into
  the `[PUBLIC_PATH]/fonts` folder.
- Bundles image assets (GIF, JPEG, JPG, PNG files) from any folder of
  source code and emits them into the `[PUBLIC_PATH]/images` folder.
- Bundle SCSS files from any folder of source code, beside `node_modules` and
  its subfolders. The files are compiled, bundled, and extracted into
  `[PUBLIC_PATH]/[CHUNK_NAME].css` chunks.
- Bundles CSS files from any folder of source code. The files are bundled and
  extracted into `[PUBLIC_PATH]/[CHUNK_NAME].css` chunks.
- Bundles JS, JSX, and SVG files. They are compiled and emitted into
  `[PUBLIC_PATH]/[CHUNK_NAME].js` chunks, using [Babel] environment
  specified in the factory options, and
  [Babel config for Webpack](/docs/api/configs/babel#client-side).
- Automatically configures the following path aliases for the compilation
  (relative paths starting with these segments are resolved relative to
  the corresponding folders):
  - **assets** for `[CONTEXT]/src/assets`;
  - **components** for `[CONTEXT]/src/shared/components`;
  - **fonts** for `[CONTEXT]/src/assets/fonts`;
  - **styles** for `[CONTEXT]/src/styles`.
- `resolve.symlinks` Webpack option is set **false** to avoid problems
  with resolution of assets from packages linked with `npm link`.
- These global variables are automatically eemulated inside the output
  JS bundle:
  - **BUILD_RNDKEY** - A random 32 bit key that can be used
    for encryption, it is set just as a global variable accessible in
    the code;
  - **BUILD_TIMESTAMP** - UTC timestamp of the beginning of the build;
  - **FRONT_END** - It is set ***true** inside the bundle, so that shared code
    can use it to determine that it is executed at the client side.
- The config opts to polyfill the `__dirname` global variable, and to ignore
  imports of the `fs` Node package;
- By default the config stores to the disk (re-writes if exists) the file
  `[CONTEXT]/.build-info` which will contain a stringified JSON
  object with the following fields:
  - **`rndkey`** &mdash; The value set for `BUILD_RNDKEY`;
  - **`timestamp`** &mdash; The value set for `BUILD_TIMESTAMP`.

### Development {#app-dev}
```js
const configFactory = require(
  '@dr.pogodin/react-utils/config/webpack/app-development',
);
```
This factory **function configFactory(options)** &rArr; **object** creates
[Webpack] configuration for **development** environment. It accepts the same
options as the [base](#app-base) app config, and does the same stuff, then on
top of that it:
- Enforces **development** [Babel] environment, and sets its `cssLocalIdent`
  argument equal `[path][name]___[local]___[hash:base64:6]` to generate verbose
  classnames for CSS modules, which are handy for debug purposes.
- Adds [`webpack-hot-middleware/client?reload=true`](https://github.com/glenjamin/webpack-hot-middleware)
  to entrypoints, which is necessary for HMR (Hot Module Reloading) setup.
- Emulates the following environment variables:
  - **BABEL_ENV** and **NODE_ENV** are both set equal **development**.
  - **DEV_TOOLS** is set equal **true**.
  - **REACT_GLOBAL_STATE_DEBUG** is set equal **true**.
- Adds the following plugins:
  - [HotModuleReplacementPlugin](https://webpack.js.org/plugins/hot-module-replacement-plugin/) and
    [react-refresh-webpack-plugin](https://github.com/pmmmwh/react-refresh-webpack-plugin)
    (these are required for HMR).

### Production {#app-prod}
```js
const configFactory = require(
  '@dr.pogodin/react-utils/config/webpack/app-production',
);
```
The factory **function configFactory(options)** &rArr; **object** creates
[Webpack] configuration for **production** environment. It accepts the same
options as the [base](#app-base) app config, and does the same stuff, then on
top of that it:
- Enforces **production** [Babel] environment.
- Emulates the following environment variables:
  - **BABEL_ENV** and **NODE_ENV** are both set equal to **production**.
- Adds the following plugins:
  - [CSS Minimizer Webpack Plugin](https://www.npmjs.com/package/css-minimizer-webpack-plugin);
  - [UglifyJsPlugin](https://webpack.js.org/plugins/uglifyjs-webpack-plugin/).

## Library
### Base {#lib-base}
```js
const configFactory = require(
  '@dr.pogodin/react-utils/config/webpack/lib-base',
);
```

The factory **function configFactory(options)** &rArr; **object** creates
base [Webpack] config for libraries, which contains the core setup common
for [development](#lib-dev) and [production](#lib-prod) library targets.
The `options` argument is an object with the following valid fields:

- **Required**:
  - `babelEnv` - **string** - [Babel] environment.
  - `context` - **string** - Base URL for resolution of relative config paths.
  - `entry` - **string | string[]** - Entry point(s). The config may prepend
    them by necessary polyfills.
  - `library` - **string** - Library name. It is important for proper resolution
    of library assets.
- **Optional**:
  - `babelLoaderOptions` - **object** - Overrides default [Babel] options for
    JSX and SVG loading.
  - `cssLocalIdent` - **string** - The template for CSS classnames generated by
    [Webpack]'s `css-loader`. It should match the corresponding setting in
    [Babel] config. Defaults to `hash:base64:6`.
  - `mode` - **string** - [Webpack mode](https://webpack.js.org/concepts/mode/).
  - `outputPath` - **string** - Build output path. Defaults `build`.

The generated config does the following:
- References to font asset (EOF, OTF, SVG, TTF, WOFF, WOFF2) files located in
  the `src/assets/fonts` folder of the library's source code are rewritten to
  to `LIBRARY_NAME/src/assets/fonts/FILENAME.FILE_EXTENSION` so that the host
  package of the library is able to find and bundle them.
- Bundle SCSS files from any folder of source code, beside `node_modules` and
  its subfolders. The files are compiled, bundled and extracted into
  the `build/{type}/style.css` chunks.
- Bundles JS, JSX, and SVG files; they are compiled into
  `build/{type}/web.bundle.js` chunks, using the [Babel] environment
  specified in the factory options, and
  [client-side Babel config](/docs/api/configs/babel#client-side).
- Sets The following path aliases (relative paths starting with these segments
  are resolved relative to the corresponding folders):
  - **assets** for `[CONTEXT]/src/assets`.
  - **components** for `[CONTEXT]/src/shared/components`.
  - **fonts** for `[CONTEXT]/src/assets/fonts`.
  - **styles** for `[CONTEXT]/src/styles`.
- Sets `resolve.symlinks` [Webpack] option to **false** to avoid problems
  with resolution of assets from packages linked with `npm link`.
- Declares the following packages as externals:
  - `@babel/runtime`
  - `@dr.pogodin/react-global-state`
  - `@dr.pogodin/react-themes`
  - `@dr.pogodin/react-utils`
  - `axios`
  - `dayjs`
  - `lodash`
  - `prop-types`
  - `qs`
  - `react`
  - `react-dom`
  - `react-helmet`
  - `react-router-dom`
  - `uuid`
  - `url-parse`

### Development {#lib-dev}
```js
const configFactory = require(
  '@dr.pogodin/react-utils/config/webpack/lib-development',
);
```
Extends and tunes the [base library config](#lib-base) for **development**
environment. This factory **function configFactory(options)** &rArr; **object**
accepts the same options, and does the same stuff. On top of that
it modifies the template for generated classnames to equal
`[path][name]___[local]___[hash:base64:6]` which is handy for debug purposes,
and does some other small modifications of config.

### Production {#lib-prod}
```js
const configFactory = require(
  '@dr.pogodin/react-utils/config/webpack/lib-production',
);
```
Extends and tunes the [base library config](#lib-base) for **production**
environment. This factory **function configFactory(options)** &rArr; **object**
accepts the same options, and does all the same stuff, and on top of that
slightly modifies the resulting config as required for production needs.

[Application]: #application
[Babel]: https://babeljs.io/
[Library]: #library
[Webpack]: https://webpack.js.org
