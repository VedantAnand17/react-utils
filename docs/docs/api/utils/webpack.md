# webpack
```jsx
import { webpack } from '@dr.pogodin/react-utils';
```
Encapsulates Webpack-related utilities.

**Functions**
- [requireWeak()](#requireweak) - Requires a module preventing it from being
  bundled into the client-side code by Webpack.
- [resolveWeak()](#resolveweak) - Resolves a module path with the help of
  Babel module resolver, thus the same way the resolution works for the normal
  `require()`.

## Functions

### requireWeak()
```jsx
webpack.requireWeak(modulePath): object
```
Require the module at `modulePath` preventing it from being bundled into
the client-side code by Webpack.

:::caution Beware
Avoid passing in relative paths: they will be resolved relative to the library
distribution folder, which is wrong in most cases. To convert a relative path
to the absolute one without breaking the client-side code due to leaking
server-side-only functions to the frontend follow this:

**Example**
```jsx
import { isomorphy, webpack } from '@dr.pogodin/react-utils';

// This block executes at the server-side only.
let serverSideOnlyModule;
if (isomorphy.IS_SERVER_SIDE) {
  // webpack.requireWeak() ensures that server-side only modules, like NodeJS
  // "path" and dummy "server-side-only-module" are not bundled in into
  // the client-side code compiled by Webpack.
  const path = webpack.requireWeak('path');

  // The first line is needed only if additional Babel module resolution is
  // needed to find the given module path.
  const relPath = webpack.resolveWeak('path/to/server-side-only-module');

  // This turns a relative path into the absolute one.
  const absPath = path.resolve(__dirname, relPath);

  serverSideOnlyModule = webpack.requireWeak(p);
}
```
:::

**Arguments & Result**
- `modulePath` - **string** - Absolute path to the module to load.
- Returns **object** - Loaded module.

### resolveWeak()
```jsx
webpack.resolveWeak(modulePath): string
```
Resolves `modulePath` module path with the help of Babel module resolver,
thus the same way the resolution works for the normal `require()`.

**Arguments & Result**
- `modulePath` - **string** - Module path.
- Returns **string** - The resolved relative path to the module.
