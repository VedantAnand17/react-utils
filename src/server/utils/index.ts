// Aggregates utility modules together.

export * as errors from './errors';
