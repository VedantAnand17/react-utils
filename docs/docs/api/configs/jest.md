# Jest
Default configuration for [Jest] testing framework.

:::caution Beware
It does strong assumptions about the host project's code structure: the host
[Jest] configuration in the `/config/jest` folder of host code; the source code
in `/src` folder; build outputs in `/build` folder; _etc._

For a project with different structure it might be easier to compose
the appropriate Jest config from scratch.
:::

To use this default configuration create two files in the `/config/jest` folder
of host project:

```js title="/config/jest/config.js"
const config = require('@dr.pogodin/react-utils/config/jest/default');

// You can tune config as necessary here.

module.exports = config;
```

```js title="/config/jest/setup.js"
import '@dr.pogodin/react-utils/config/jest/setup';
```

After that point to that host config file from the Jest execution command,
_e.g._:
```bash
jest --config config/jest/config.js
```

[Jest]: https://jestjs.io
