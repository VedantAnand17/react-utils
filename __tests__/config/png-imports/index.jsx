import path from 'path';

import webpack, { ProgressPlugin } from 'webpack';

import { snapshot } from 'utils/jest';

import webpackConfigFactoryAppDev
  from '../../../config/webpack/app-production';

import App from './__assets__';

function testWebpackBuild(configFactory) {
  const config = configFactory({
    context: process.cwd(),
    dontEmitBuildInfo: true,
  });
  config.plugins = config.plugins.filter((p) => !(p instanceof ProgressPlugin));
  config.entry = './__tests__/config/png-imports/__assets__';
  const babelLoader = config.module.rules.find(
    (x) => x.loader === 'babel-loader',
  );
  babelLoader.options.presets[0] = path.resolve('config/babel/webpack');
  const compiler = webpack(config);
  compiler.hooks.shouldEmit.tap('Test', () => false);
  return new Promise((resolve) => {
    compiler.run((err, stats) => {
      expect(err).toBe(null);
      expect(stats.compilation.errors).toEqual([]);
      const pngChunkName = Object.keys(stats.compilation.assets)
        .find((key) => key.endsWith('.png'));
      expect(pngChunkName).toMatchSnapshot();
      resolve(pngChunkName);
    });
  });
}

test('Client-side import', async () => {
  await testWebpackBuild(webpackConfigFactoryAppDev);
}, 30000);

test('Server-side import', () => {
  // NOTE: The PNG asset name in the snapshot must match that generated
  // in Webpack builds.
  snapshot(<App />);
});
