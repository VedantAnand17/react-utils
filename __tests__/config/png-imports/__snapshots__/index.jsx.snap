// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client-side import 1`] = `"images/cd564973d92d0a8a8a17.png"`;

exports[`Server-side import 1`] = `
<img
  alt="logo"
  src="/images/cd564973d92d0a8a8a17.png"
/>
`;
